File: implementations/csharp/Cid.cs
using System.Security.Cryptography;

namespace Implementations.CSharp;

internal static class Cid
{
    public static string BaseDirectory { get; } = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", ".."));

    public static string ExamplesDirectory => Path.Combine(BaseDirectory, "examples");

    public static string CidsDirectory => Path.Combine(BaseDirectory, "cids");

    public static string Compute(byte[] content)
    {
        var prefix = EncodeLength(content.Length);
        var suffix = content.Length <= 64 ? ToBase64Url(content) : ToBase64Url(Hash(content));
        return prefix + suffix;
    }

    private static string EncodeLength(int length)
    {
        var buffer = new byte[6];
        var remaining = (ulong)length;

        for (var i = 0; i < buffer.Length; i++)
        {
            buffer[buffer.Length - 1 - i] = (byte)(remaining >> (8 * i));
        }
        return ToBase64Url(buffer);
    }

    private static byte[] Hash(byte[] content)
    {
        using var sha = SHA512.Create();
        return sha.ComputeHash(content);
    }

    private static string ToBase64Url(byte[] data)
    {
        var encoded = Convert.ToBase64String(data);
        return encoded.TrimEnd('=').Replace('+', '-').Replace('/', '_');
    }

    public record DownloadResult(byte[] Content, string Computed, bool IsValid);

    public static async Task<DownloadResult> DownloadAsync(string baseUrl, string cid)
    {
        var url = baseUrl.TrimEnd('/') + '/' + cid;
        using var client = new HttpClient();
        client.Timeout = TimeSpan.FromSeconds(10);
        
        var response = await client.GetAsync(url);
        if (!response.IsSuccessStatusCode)
        {
            throw new HttpRequestException($"HTTP {(int)response.StatusCode}: {response.ReasonPhrase}");
        }
        
        var content = await response.Content.ReadAsByteArrayAsync();
        var computed = Compute(content);
        var isValid = computed == cid;
        
        return new DownloadResult(content, computed, isValid);
    }
}

File: implementations/dart/cid.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

final Directory _scriptDir = File(Platform.script.toFilePath()).parent;
final Directory repoDir = _scriptDir.parent.parent;
final Directory cidsDir = Directory('${repoDir.path}/cids');

String _toBase64Url(Uint8List data) {
  return base64UrlEncode(data).replaceAll('=', '');
}

String encodeLength(int length) {
  final buffer = Uint8List(6);
  var value = length;
  for (var i = 5; i >= 0; i--) {
    buffer[i] = value & 0xff;
    value >>= 8;
  }
  return _toBase64Url(buffer);
}

String computeCid(List<int> content) {
  final prefix = encodeLength(content.length);
  final suffix = content.length <= 64
      ? _toBase64Url(Uint8List.fromList(content))
      : _toBase64Url(Uint8List.fromList(sha512.convert(content).bytes));
  return '$prefix$suffix';
}

class DownloadResult {
  final Uint8List content;
  final String computed;
  final bool isValid;

  DownloadResult(this.content, this.computed, this.isValid);
}

Future<DownloadResult> downloadCid(String baseUrl, String cid) async {
  final url = '${baseUrl.replaceAll(RegExp(r'/$'), '')}/$cid';
  final client = HttpClient();
  
  try {
    final request = await client.getUrl(Uri.parse(url));
    final response = await request.close();
    
    if (response.statusCode != 200) {
      throw Exception('HTTP ${response.statusCode}: ${response.reasonPhrase}');
    }
    
    final bytes = await response.expand((chunk) => chunk).toList();
    final content = Uint8List.fromList(bytes);
    final computed = computeCid(content);
    final isValid = computed == cid;
    
    return DownloadResult(content, computed, isValid);
  } finally {
    client.close();
  }
}
